### PR TITLE
Dev 7020 markup include

### PR DESCRIPTION
**Pull request by Matt-DiPaolo**
Testing Scenario setup:

- User is in COVE Editor
- Select Dashboard Visualization
- Select any data source
- Drag and drop the markup include to the visualization panel
- Either configure the data source with the gear icon or do not use any data source. 
- Click to tool icon to bring up the table editor panel.

Testing instructions:

1. Edit the Header
- Open the General Accordion on the left.
- The header title can be changed from the text input.
- If the Header is left blank, then the header section will not be shown in the Markup Include.

2. Add HTML Markup:
- Open the Content Editor accordion on the left.
- The "USE INLINE HTML" checkbox with toggle between using the HTML markup and the legacy URL markup.
- Within the HTML text box, adding basic HTML elements will show up on the Markup Include section.
- Bootstrap classes and inline styling will work in this section.
- Variables can be added here using the following pattern {{variable name}}.
   - example: Variable name of State would be written out as {{State}}.
   - see next step for adding Variables.
  
3. Adding variables
- To use Variables, a data source needs to be imported and configured.
- Below the HTML Text Box, there is the Variable section.
- Click the "Create New Variable" button and a "New Variable" with a down arrow button will appear.
- Clicking on the arrow will open up the new variable.
- Add a name to the "VARIABLE NAME" input. The name can be renamed later if needed.
- Once the name is inputted, the "COLUMN" dropdown with be available.
- Select a Column. This will be the output of the variable.
  - example: If there is a State column, the variable will display the state that meets the conditions set by global page filter and the optional conditions that will be set below.

4. Setting conditions 
- After the variable name and column are selected, the "ADD CONDITION" button will be available. 
- Conditions are optional and can add as many as needed.
- Click the "Add Condition" button and a "NEW CONDITION" with a down arrow button will appear.
- Clicking on the arrow will open up the new condition.
- The first select will set the condition to validate
- The second select will set if the condition is or is not equal to the value
- The third select will set the value of the condition
- After the "Add Condition" can be pushed again or "Done" if no more conditions are needed

5. Editing and deleting variables
- Click on the down arrow of the Variable that needs to be edited or deleted
- To edit the variable, just change any field that was set. 
- If "COLUMN" is changed, this will delete the conditions that were set earlier for that variable because conditions are based off the column. 
- Press the "Delete" button to delete the variable

6. Editing and deleting conditions
- Click on the down arrow of the Condition that needs to be edited or deleted
- To edit the condition, just change any select that was set. 
- If the first select is changed, this will reset the third select  
- Press the "Remove" button to delete the condition

7. Using variables
- As stated in step 2, to add a variable to the Markup Include, wrap the variable in {{variable name}}.
- When a variable is used, in the Markup, the variable will be highlighted yellow. 
- Hover over the variable to see the possible values that will replace the variable on a live page.
- These are possible values that will be determine by the global filter on the page.
- The same variable can be added to the HTML as many times as needed.

### CODE REVIEW

File: packages/markup-include/src/components/EditorPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+      }
+    }
+
+    const { currentConditionColumnName, currentConditionValueList, currentConditionNumber, currentIsOrIsNotEqualTo, currentValue } =
+      currentConditionSettings
+
+    const handleConditionChange = (selectionValue, conditionSetting) => {
+      const conditionSettingUpdate = conditionChangeSwitch({ selectionValue, conditionSetting })
+      setNewCurrentConditionSettings(conditionSettingUpdate)
+      updateConditionsList(conditionSettingUpdate)
```
Comment (joshlacey) [inline packages/markup-include/src/components/EditorPanel.tsx:174]: new changes 174 - 558

File: packages/markup-include/src/components/EditorPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+  const handleAddConditionClick = () => {
+    const newConditionsList = [..._.cloneDeep(conditionsList)]
+    newConditionsList.push({
+      columnName: '',
+      isOrIsNotEqualTo: 'is',
+      value: ''
+    })
+    setConditionsList(newConditionsList)
+    const [conditionControls, setConditionControls] = openConditionControls
+    setConditionControls({ ..._.cloneDeep(conditionControls), [conditionsList.length + 1]: true })
```
Comment (joshlacey) [inline packages/markup-include/src/components/EditorPanel.tsx:180]: don't deep clone here

Comment (joshlacey) [reply packages/markup-include/src/components/EditorPanel.tsx:180]: Also it looks like you're using a Record as though it were an array?

File: packages/markup-include/src/components/EditorPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+  fieldName: string
+  label: string
+  section?: string
+  subsection?: string
+  tooltip?: string
+  updateField: Function
+  value: boolean
+}
+
+const CheckBox: React.FC<CheckBoxProps> = memo(({ fieldName, label, subsection = null, tooltip, section = null, updateField, value, ...attributes }) => (
```
Comment (joshlacey) [inline packages/markup-include/src/components/EditorPanel.tsx:285]: can we refactor and use this checkbox? \packages\core\components\inputs\InputCheckbox.jsx

File: packages/markup-include/src/components/EditorPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+  const handleCreateNewVariableButtonClick = () => {
+    const newVariableArray = [..._.cloneDeep(variableArray)]
+    const newVariable = {
+      columnName: '',
+      conditions: [],
+      name: '',
+      tag: ''
+    }
+
+    setControls({ ...controls, [variableArray.length + 1]: true })
```
Comment (joshlacey) [inline packages/markup-include/src/components/EditorPanel.tsx:392]: why is this being used like an array?

File: packages/markup-include/src/components/EditorPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+
+  const convertStateToConfig = () => {
+    let strippedState = JSON.parse(JSON.stringify(config))
+    delete strippedState.newViz
+    delete strippedState.runtime
+
+    return strippedState
+  }
+
+  const [variableArray, setVariableArray] = useState<Variable[]>([])
```
Comment (joshlacey) [inline packages/markup-include/src/components/EditorPanel.tsx:376]: I thought we moved all this up to the parent config? are we sure this is getting saved?

File: packages/markup-include/src/components/EditorPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+    const newVariableArray = [..._.cloneDeep(variableArray)]
+    newVariableArray[variableIndex] = newVariable
+    setVariableArray(newVariableArray)
+    updateField('contentEditor', null, 'markupVariables', newVariableArray)
+    config.contentEditor.markupVariables
+    return
+  }
+
+  const deleteVariable = (variableIndex: number) => {
+    const newVariableArray = [..._.cloneDeep(variableArray)]
```
Comment (joshlacey) [inline packages/markup-include/src/components/EditorPanel.tsx:409]: fix

File: packages/markup-include/src/components/EditorPanel.tsx [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+    newVariableArray.splice(variableIndex, 1)
+    setVariableArray(newVariableArray)
+    updateField('contentEditor', null, 'markupVariables', newVariableArray)
+    config.contentEditor.markupVariables
+
+    const newControls = { ..._.cloneDeep(controls) }
```
Comment (joshlacey) [inline packages/markup-include/src/components/EditorPanel.tsx:415]: fix

File: packages/markup-include/src/data/markup-include-example.json [hunk 8]
```diff
@@ -0,0 +1,362 @@
+[
```
Comment (joshlacey) [inline packages/markup-include/src/data/markup-include-example.json:1]: I don't think we need any addititional data sources. we should be able to use the ones we've already stored

File: packages/markup-include/src/data/sample-content-editor.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
+export default {
```
Comment (joshlacey) [inline packages/markup-include/src/data/sample-content-editor.tsx:1]: what's this?

**Comment by joshlacey** [comment]
looks like you have an empty file here: packages/markup-include/src/columnsEditor.tsx

File: packages/markup-include/src/CdcMarkupInclude.tsx [hunk 10]
```diff
@@ -132,7 +133,45 @@ const CdcMarkupInclude = (props: CdcMarkupIncludeProps) => {
...
+  const filterOutConditions = (workingData, conditionList) => {
+    const { columnName, isOrIsNotEqualTo, value } = conditionList[0]
+
+    const newWorkingData = isOrIsNotEqualTo === 'is' ? workingData.filter(dataObject => dataObject[columnName] === value) : workingData.filter(dataObject => dataObject[columnName] !== value)
+
+    conditionList.shift()
+    return conditionList.length === 0 ? newWorkingData : filterOutConditions(newWorkingData, conditionList)
+  }
+
+  const convertVariablesInMarkup = inlineMarkup => {
```
[the remark this answers is not in the document]
Comment (Matt-DiPaolo) [reply packages/markup-include/src/CdcMarkupInclude.tsx:147]: would like to talk this one over with you to move forward.

File: packages/markup-include/src/components/EditorPanel.tsx (hunk 2 above)
Comment (Matt-DiPaolo) [reply packages/markup-include/src/components/EditorPanel.tsx:180]: Removed deep clone but, need to talk through the Record as an array.

File: packages/markup-include/examples/example-config.json [hunk 11]
```diff
@@ -1,11 +1,10 @@
 {
-  "type": "markup-include",
-  "srcUrl": "#example",
-  "title": "Markup Include",
-  "theme": "theme-blue",
-  "data": [],
-  "runtime": {
-    "editorErrorMessage": "Test"
+  "contentEditor": {
```
Comment (joshlacey) [inline packages/markup-include/examples/example-config.json:2]: Please delete this file

File: packages/markup-include/src/data/initial-state.tsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
+export default {
+  type: 'data-bite',
```
Comment (joshlacey) [inline packages/markup-include/src/data/initial-state.tsx:2]: what is this data-bite initial state??

File: packages/markup-include/index.html [hunk 13]
```diff
@@ -22,10 +22,9 @@
   </head>
   <body>
     <!-- Original -->
-    <div class="react-container" data-config="/examples/example-config.json"></div>
 
     <!-- DATA PRESENTATION GALLERY: https://www.cdc.gov/wcms/4.0/cdc-wp/data-presentation/Markup-Include.html#examples -->
-    <!-- <div class="react-container" data-config="/examples/gallery/icon-no-text.json"></div> -->
+    <div class="react-container" data-config="/examples/gallery/icon-no-text.json"></div>
```
Comment (joshlacey) [inline packages/markup-include/index.html:27]: please remove changes to this file

File: packages/markup-include/src/data/markup-include-example.json (hunk 8 above)
Comment (joshlacey) [inline packages/markup-include/src/data/markup-include-example.json:1]: delete this

**Comment by adamdoe** [comment]
@Matt-DiPaolo no rush here, just curious on the status. I'm unable to run the testing files today.

File: packages/core/package.json [hunk 14]
```diff
@@ -6,7 +6,8 @@
   "main": "dist/cdccore",
   "scripts": {
     "test": "vitest watch --reporter verbose --globals true",
-    "test:ui": "vitest --ui"
+    "test:ui": "vitest --ui",
+    "build" : ""
```
Comment (joshlacey) [inline packages/core/package.json:10]: did this work?

File: packages/dashboard/src/CdcDashboardComponent.tsx [hunk 15]
```diff
@@ -24,7 +24,7 @@ import CdcMap from '@cdc/map'
 import CdcChart from '@cdc/chart'
 import CdcDataBite from '@cdc/data-bite'
 import CdcWaffleChart from '@cdc/waffle-chart'
-import CdcMarkupInclude from '@cdc/markup-include'
+import CdcMarkupInclude from '@cdc/markup-include/src/CdcMarkupInclude'
```
Comment (joshlacey) [inline packages/dashboard/src/CdcDashboardComponent.tsx:27]: change this back to @cdc/markup-include

File: packages/dashboard/src/CdcDashboardComponent.tsx [hunk 16]
```diff
@@ -765,7 +776,7 @@ export default function CdcDashboard({ initialState, isEditor = false, isDebug =
 
   return (
     <GlobalContextProvider>
-      <DashboardContext.Provider value={{ ...state, setParentConfig: editorContext.setTempConfig, outerContainerRef, isDebug }}>
+      <DashboardContext.Provider value={{ ...state, setParentConfig: editorContext.setTempConfig, outerContainerRef, isDebug, isPreview }}>
```
Comment (joshlacey) [inline packages/dashboard/src/CdcDashboardComponent.tsx:779]: if you put isPreview here you don't need to also pass it as a prop on 691. VisualizationRow already is hooked into the DashboardContext. so pull it from there.

File: packages/markup-include/src/ConfigContext.tsx [hunk 17]
```diff
@@ -0,0 +1,16 @@
+import { createContext } from 'react'
```
Comment (joshlacey) [inline packages/markup-include/src/ConfigContext.tsx:1]: This should be a .ts file instead of tsx

**Review by joshlacey — CHANGES_REQUESTED** [review]

File: packages/core/helpers/ver/4.24.4.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
+import _ from 'lodash'
```
Comment (joshlacey) [inline packages/core/helpers/ver/4.24.4.ts:1]: @adamdoe should this be 4.24.5 now?

File: lerna.json [hunk 19]
```diff
@@ -1,5 +1,15 @@
 {
   "$schema": "node_modules/lerna/schemas/lerna-schema.json",
   "npmClient": "yarn",
+<<<<<<< HEAD
```
Comment (joshlacey) [inline lerna.json:4]: @Matt-DiPaolo you have merge conflicts that weren't resolved

File: packages/markup-include/src/_stories/MarkupInclude.stories.tsx [hunk 20]
```diff
@@ -1,18 +1,316 @@
...
+    config: {
+      contentEditor: {
+        inlineHTML: '<div>{{state}} does have a rate 130 compared to the over all rate of {{overall-rate}}</div>',
+        markupVariables: [],
+        showHeader: false,
+        srcUrl: 'https://www.cdc.gov/wcms/4.0/cdc-wp/data-presentation/examples/Markup-Include-Button-and-Text.html',
+        title: '',
+        useInlineHTML: true
+      },
+      data: [
```
Comment (joshlacey) [inline packages/markup-include/src/_stories/MarkupInclude.stories.tsx:176]: make data one variable or json file and import it so we don't end up with multiple duplicates of multiline data blocks

File: packages/markup-include/src/_stories/MarkupInclude.stories.tsx [hunk 21]
```diff
@@ -1,18 +1,316 @@
...
+        hideBackgroundColor: false,
+        borderColorTheme: false
+      }
+    },
+    isEditor: true,
+    showEditorPanel: true
+  }
+}
+
+export const icon_no_text: Story = {
```
Comment (joshlacey) [inline packages/markup-include/src/_stories/MarkupInclude.stories.tsx:216]: All these examples that were migrated should be imported and used directly as config

**Review by joshlacey — CHANGES_REQUESTED** [review]

File: packages/core/helpers/ver/4.24.4.ts (hunk 18 above)
Comment (adamdoe) [reply packages/core/helpers/ver/4.24.4.ts:1]: Yes, any new changes should go into 4.24.5.

Comment (joshlacey) [reply packages/core/helpers/ver/4.24.4.ts:1]: @Matt-DiPaolo can you make that change?

**Comment by joshlacey** [comment]
@Matt-DiPaolo I think you need to mark me for rereview for my "request changes" thing to go away?

**Review by joshlacey — APPROVED** [review]

**Pull request merged**